cli/src/utils/command.ts

```diff
@@ -0,0 +1,120 @@
+import { readFileSync } from 'fs';
+import { resolve } from 'path';
+import { URL } from 'url';
+import { Command } from '@oclif/command';
+import { CLIError } from '@oclif/errors';
+import * as Parser from '@oclif/parser';
+import open from 'open';
+import ms from 'ms';
+import terminalLink from 'terminal-link';
+import { LycanClient } from '@binaris/spice-node-client';
+import flags from './cli-flags';
+import userConfig from './user-config';
+
+const pjson = JSON.parse(readFileSync(resolve(__dirname, '../../package.json'), 'utf-8'));
+
+const LOGIN_PARAM = 'ticket';
+const TICKET_CLAIM_INTERVAL_MS = 1000;
+
+export default abstract class BaseCommand extends Command {
+  static cliBinName = pjson.oclif.bin as string;
+
+  private apiEndpoint?: string;
+  private webAppEndpoint?: string;
+  protected _lycanClient?: LycanClient;
+  protected get lycanClient(): LycanClient {
+    if (!this._lycanClient) {
+      this._lycanClient = this.createLycanClient(userConfig.get('accessToken') as string | undefined);
+    }
+    return this._lycanClient;
+  }
+
+  public static flags = {
+    help: flags.help({ char: 'h' }),
+    apiEndpoint: flags.string({
+      default: 'https://api.shiftjs.com',
+      hidden: true,
+      env: 'SHIFTJS_API_ENDPOINT',
+    }),
+    webAppEndpoint: flags.string({
+      default: 'https://app.shiftjs.com',
+      hidden: true,
+      env: 'SHIFTJS_WEBAPP_ENDPOINT',
+    }),
+  };
+
+  protected parse<F, A extends {[name: string]: any}>(opt?: Parser.Input<F>, argv = this.argv): Parser.Output<F, A> {
+    if (!opt) {
+      opt = this.constructor as any;
+    }
+    if (!opt!.flags) {
```

Comment:
Do we still need the `!`?

Reply:
yes, but I think it never happens, I can add a check/throw if you want

Reply:
changed how the typing here is handled